### PR TITLE
Add service methods for immutable Compute resources and tests

### PR DESCRIPTION
**Pull request by mziccard**
This PR adds spi and service methods for manipulating the following resources:
- DiskType
- MachineType
- License
- Zone
- Region

In c18730c integration tests are also added.

### CODE REVIEW

**Comment by mziccard** [comment]
I think the coverage drop is the usual one we see when we add the spi layer that is only tested through ITs (which do not concur to compute the coverage).

File: gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java [hunk 1]
```diff
@@ -43,4 +64,214 @@ public DefaultComputeRpc(ComputeOptions options) {
...
+          .aggregatedList(this.options.projectId())
+          .setFilter(FILTER.getString(options))
+          .setMaxResults(MAX_RESULTS.getLong(options))
+          .setPageToken(PAGE_TOKEN.getString(options))
+          .setFields(FIELDS.getString(options))
+          .execute();
+      ImmutableList.Builder<DiskType> builder = ImmutableList.builder();
+      Map<String, DiskTypesScopedList> scopedList = aggregatedList.getItems();
+      if (scopedList != null) {
+        for (String key : scopedList.keySet()) {
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java]: entrySet?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java [hunk 2]
```diff
@@ -24,4 +30,548 @@
...
+    VALID_DISK_SIZE("validDiskSize"),
+    ZONE("zone");
+
+    private final String selector;
+
+    DiskTypeField(String selector) {
+      this.selector = selector;
+    }
+
+    public String selector() {
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: I am not sure if we discussed that already.
If we don't mind to make this method `public` than we could make all the XXXField enums implement a "Selector" interface and have one static `String selector(Selector required, Selector... other)` method.

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java [hunk 3]
```diff
@@ -24,4 +30,548 @@
...
+  /**
+   * Fields of a Compute Engine DiskType resource.
+   *
+   * @see <a href="https://cloud.google.com/compute/docs/reference/latest/diskTypes#resource">Disk
+   *     Type Resource</a>
+   */
+  enum DiskTypeField {
+    CREATION_TIMESTAMP("creationTimestamp"),
+    DEFAULT_DISK_SIZE_GB("defaultDiskSizeGb"),
+    DESCRIPTION("description"),
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: did you intentionally left out "deprecated"?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java (hunk 2 above)
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: Yes we discussed it and opened #456 but it seems like we haven't had time to address that issue.

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java [hunk 4]
```diff
@@ -24,4 +30,548 @@
...
+  /**
+   * Fields of a Compute Engine MachineType resource.
+   *
+   * @see <a href="https://cloud.google.com/compute/docs/reference/latest/machineTypes#resource">
+   *     Machine Type Resource</a>
+   */
+  enum MachineTypeField {
+    CREATION_TIMESTAMP("creationTimestamp"),
+    DESCRIPTION("description"),
+    GUEST_CPUS("cpus"),
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: "cpus" -> "guestCpus"?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java [hunk 5]
```diff
@@ -24,4 +30,548 @@
...
+    ID("id"),
+    IMAGE_SPACE_GB("imageSpaceGb"),
+    MAXIMUM_PERSISTENT_DISKS("maximumPersistentDisks"),
+    MAXIMUM_PERSISTENT_DISKS_SIZE_GB("maximumPersistentDisksSizeGb"),
+    MEMORY_MB("memoryMb"),
+    NAME("name"),
+    SCRATCH_DISKS("scratchDisks"),
+    DISK_GB("diskGb"),
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: isn't that part of "scratchDisks"?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java [hunk 6]
```diff
@@ -24,4 +30,548 @@
...
+     */
+    public static DiskTypeOption fields(DiskTypeField... fields) {
+      return new DiskTypeOption(ComputeRpc.Option.FIELDS, DiskTypeField.selector(fields));
+    }
+  }
+
+  /**
+   * Class for specifying disk type list options.
+   */
+  class DiskTypeListOption extends Option {
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: filter is coming later?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java [hunk 7]
```diff
@@ -24,4 +30,548 @@
...
+
+    /**
+     * Returns an option to specify the region's fields to be returned by the RPC call. If this
+     * option is not provided all region's fields are returned. {@code RegionListOption.fields} can
+     * be used to specify only the fields of interest. {@link Region#regionId()} is always
+     * returned, even if not specified.
+     */
+    public static RegionListOption fields(RegionField... fields) {
+      StringBuilder builder = new StringBuilder();
+      builder.append("items(").append(RegionField.selector(fields)).append("),nextPageToken");
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: With the "Selectable" interface we could also make fields -> list string a common method...

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java [hunk 8]
```diff
@@ -24,4 +30,548 @@
...
+      return new LicenseOption(ComputeRpc.Option.FIELDS, LicenseField.selector(fields));
+    }
+  }
+
+  /**
+   * Returns the requested disk type or {@code null} if not found.
+   *
+   * @throws ComputeException upon failure
+   */
+  DiskType getDiskType(DiskTypeId diskTypeId, DiskTypeOption... options) throws ComputeException;
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: No need for "throws ComputeException". Having it in the javadoc is enough.

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java [hunk 9]
```diff
@@ -24,4 +30,548 @@
...
+  /**
+   * Returns the requested disk type or {@code null} if not found.
+   *
+   * @throws ComputeException upon failure
+   */
+  DiskType getDiskType(String zone, String diskType, DiskTypeOption... options)
+      throws ComputeException;
+
+  /**
+   * Lists the disk types in the provided zone available to the current project.
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: is the "to the current project" necessary or inferred?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java [hunk 10]
```diff
@@ -16,15 +16,469 @@
...
+      ComputeRpc.Tuple<String, Iterable<com.google.api.services.compute.model.DiskType>> result =
+          runWithRetries(new Callable<ComputeRpc.Tuple<String,
+              Iterable<com.google.api.services.compute.model.DiskType>>>() {
+            @Override
+            public ComputeRpc.Tuple<String,
+                Iterable<com.google.api.services.compute.model.DiskType>> call() {
+              return serviceOptions.rpc().listDiskTypes(optionsMap);
+            }
+          }, serviceOptions.retryParams(), EXCEPTION_HANDLER);
+      String cursor = result.x();
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java]: Any reason for including it inside of the retry handling try~catch?

File: gcloud-java-compute/src/main/java/com/google/gcloud/spi/ComputeRpc.java [hunk 11]
```diff
@@ -16,5 +16,155 @@
...
+      return y;
+    }
+  }
+
+  /**
+   * Returns the requested disk type or {@code null} if not found.
+   *
+   * @throws ComputeException upon failure
+   */
+  DiskType getDiskType(String zone, String diskType, Map<Option, ?> options)
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/spi/ComputeRpc.java]: same comment about throws.

File: gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java [hunk 12]
```diff
@@ -43,4 +64,214 @@ public DefaultComputeRpc(ComputeOptions options) {
...
+  @Override
+  public MachineType getMachineType(String zone, String machineType, Map<Option, ?> options)
+      throws ComputeException {
+    try {
+      return compute.machineTypes()
+          .get(this.options.projectId(), zone, machineType)
+          .setFields(FIELDS.getString(options))
+          .execute();
+    } catch (IOException ex) {
+      ComputeException serviceException = translate(ex);
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java]: if you wanted you can make this re-occuring code a generic method that returns T (always null) or throws ComputeException.

File: gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java [hunk 13]
```diff
@@ -43,4 +64,214 @@ public DefaultComputeRpc(ComputeOptions options) {
...
+    }
+  }
+
+  @Override
+  public Tuple<String, Iterable<DiskType>> listDiskTypes(String zone, Map<Option, ?> options)
+      throws ComputeException {
+    try {
+      DiskTypeList diskTypesList = compute.diskTypes()
+          .list(this.options.projectId(), zone)
+          .setFilter(FILTER.getString(options))
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java]: Not used yet, right?

File: gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java [hunk 14]
```diff
@@ -43,4 +64,214 @@ public DefaultComputeRpc(ComputeOptions options) {
   private static ComputeException translate(IOException exception) {
     return new ComputeException(exception);
   }
+
+  @Override
+  public DiskType getDiskType(String zone, String diskType, Map<Option, ?> options)
+      throws ComputeException {
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java]: No need for the explicit "throws ComputeException" (I see that in Storage we do that in the spi RPC interface and implementation but I think it is a leftover as originally it was IOException). In any case, for consistency I am fine with leaving it as is (here and in the spi interface), but it should be removed from the API interface (that has a javadoc to cover it).

**Comment by aozarov** [comment]
Nice work! Done with first iteration. Any reason we don't have codacy analysis here?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java (hunk 10 above)
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java]: No particular reason but this method has the same structure we used almost everywhere else for listing methods.

**Comment by mziccard** [comment]
@aozarov I addressed the comments in 28ed6e8

- Add support for filters: `filter` method to options + filter classes
- Removed unnecessary fields + added missing ones
- Removed redundant "throws" from method signatures
- Add static method in `DefaultComputeRpc` to handle the NOT_FOUND case

I also had to add dedicated option classes for aggregated lists as `fields` option is not completely supported (see #711).

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java [hunk 15]
```diff
@@ -24,4 +35,858 @@
...
+   */
+  abstract class ListFilter implements Serializable {
+
+    private static final long serialVersionUID = -238638392811165127L;
+
+    private final String field;
+    private final ComparisonOperator operator;
+    private final Object value;
+
+    enum ComparisonOperator {
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: public

Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: Oh, looks like you meant it to be package scoped. Fine in that case.

Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: The Beta API also supports filtering on multiple expressions. Is it something that you plan to add later?

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java (hunk 10 above)
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java]: Ok, Fine as is, but I don't think we need to consider that as a recommended way or a pattern we should follow (nor should we go and change other places).

File: gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java [hunk 16]
```diff
@@ -43,4 +64,201 @@ public DefaultComputeRpc(ComputeOptions options) {
...
+
+  @Override
+  public Tuple<String, Iterable<MachineType>> listMachineTypes(Map<Option, ?> options) {
+    try {
+      MachineTypeAggregatedList aggregatedList = compute.machineTypes()
+          .aggregatedList(this.options.projectId())
+          .setFilter(FILTER.getString(options))
+          .setMaxResults(MAX_RESULTS.getLong(options))
+          .setPageToken(PAGE_TOKEN.getString(options))
+          // todo(mziccard): uncomment or remove once #711 is closed
```
Comment (aozarov) [inline gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java]: unnecessary comment?

**Comment by aozarov** [comment]
Looks good. Few minor comments. Feel free to merge after they are resolved.

File: gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java (hunk 15 above)
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java]: Yes we decided to stick to v1, I plan to support multiple filters as soon as the Beta API becomes v2. This should be quite simple.

File: gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java (hunk 16 above)
Comment (mziccard) [inline gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java]: I forgot to comment the following line.

**Pull request merged**